proposals/0010-resource-element-type-validation.md

```diff
@@ -0,0 +1,160 @@
+* Proposal: [0010](0010-resource-element-type-validation.md)
+* Author(s): [Joshua Batista](https://github.com/bob80905)
+* Sponsor: Joshua Batista
+* Status: **Under Consideration**
+* Impacted Project(s): (LLVM)
+
+* Issues: [#75676](https://github.com/llvm/llvm-project/issues/75676)
+
+## Introduction
+Resources are often used in HLSL, with various resource element types (RETs).
+For example:
+```
+RWBuffer<float> rwbuf: register(u0);
+```
+In this code, the RET is `float`, and the resource type is `RWBuffer`.
+There are two types of buffers, RawBuffers and TypedBuffers. `RWBuffer`
+is a TypedBuffer variant, and `StructuredBuffer` is a RawBuffer variant.
+There is a distinct set of rules that define valid RETs for RawBuffer types, 
+and a separate set of rules that define valid RETs for TypedBuffer types.
+These rules also depend on the target IR, SPIR-V or DXIL.
+
+RETs for TypedBuffer variants may include:
+* basic types: 
+  * 16- and 32-bit int and uint
+  * half and float
+* vectors and matrices 
+  * containing 4 elements or fewer
+  * total size may not exceed 128 bits
+* user defined types (structs / classes), as long as:
+  * all fields in the struct have the same type
+  * there are at most 4 sub elements
+  * total size may not exceed 128 bits
+
+RETs for RawBuffer variants are much less constrained:
+* it must be a complete type
+* cannot contain a handle or resource type

```

Comment:
Is this list complete? Are there going to be intangible types that aren't handles?

Reply:
I am unsure about our plan for the implementation of RayQuery, it will be intangible but will not be a resource. So, it may not have a handle. It might be feasible that there will exist intangible types that aren't handles.

Reply:
So would it be better to say it cannot contain an intangible type?